# Chapter 5 : System Response: Evaluating the Effectiveness of Fraud Detection

## Introduction
This chapter evaluates how effectively the system detects and responds to fraudulent activity in the mobile money ecosystem. It focuses on the speed, accuracy, and overall performance of fraud detection mechanisms in limiting financial losses and disrupting fraudulent transactions. The analysis assesses whether current controls are sufficient to prevent fraud from fully materializing or spreading across the system. When detection systems fail to catch fraud in real time, losses are fully realized by victims before any intervention can occur, reducing recovery possibilities and allowing fraudulent activity to continue undetected within the ecosystem.

### Contents
- [Q1: Overall Detection Effectiveness](#q1)
- [Q2: Scale of Undetected Fraud](#q2)
- [Q3: False Positive Rate and User Experience Impact](#q3)

## Database Connection

In [1]:
# Establish connection to the MySQL database using the custom utility module
# Returns:
# - engine: used for running SQL queries
# - conn: active database connection
from db_connect import connect_to_db
engine, conn = connect_to_db()

Please enter database name:  fraud_analysis


   Using default: fraud_analysis


 Please enter the password for root@127.0.0.1:  ········



 The notebook is successfully connected to MYSQL!
  Server: 127.0.0.1:3306
  Database in use: fraud_analysis
  User: root
SQL magic configured - use %%sql in any cell!



## Q1: Overall Detection Effectiveness
How effectively does the fraud detection system identify fraudulent transactions overall, and what does the balance between detected and undetected fraud reveal about the system's reliability in protecting the platform?

In [4]:
%%sql
WITH classification AS (
    SELECT
        CASE
            WHEN isFraud = 1 AND isFlaggedFraud = 1 THEN 'Caught (True Positive)'
            WHEN isFraud = 1 AND isFlaggedFraud = 0 THEN 'Missed (False Negative)'
            WHEN isFraud = 0 AND isFlaggedFraud = 1 THEN 'False Alarm (False Positive)'
            WHEN isFraud = 0 AND isFlaggedFraud = 0 THEN 'Correct (True Negative)'
        END AS category
    FROM paysim
)

SELECT
    category,
    COUNT(*) AS transaction_count,
    CONCAT(ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM paysim), 4), '%') AS percentage
FROM classification
GROUP BY category
ORDER BY transaction_count DESC;

 * mysql+pymysql://root:***@127.0.0.1:3306/fraud_analysis
3 rows affected.


category,transaction_count,percentage
Correct (True Negative),6354407,99.8709%
Missed (False Negative),8197,0.1288%
Caught (True Positive),16,0.0003%


### Insight 
The fraud detection system performs poorly in identifying fraud despite a high overall accuracy of 99.87%, which is misleading due to class imbalance. The system detects only 16 fraudulent cases while missing 8,197, showing that it fails to capture almost all actual fraud activity. The absence of false positives suggests the detection mechanism is highly conservative, prioritizing certainty over coverage and resulting in extremely low recall. From a criminological perspective, this low detection rate significantly reduces perceived risk for offenders, weakening deterrence and allowing fraudulent behavior to persist with minimal constraint.

## Q2: Scale of Undetected Fraud
What proportion and scale of actual fraudulent transactions are not detected by the system, and what does this reveal about the level of exposure the platform has to undetected financial crime?

In [5]:
%%sql
WITH missed_fraud AS (
    SELECT
        (oldbalanceOrg - newbalanceOrig) AS loss
    FROM paysim
    WHERE isFraud = 1
      AND isFlaggedFraud = 0
)

SELECT
    COUNT(*) AS total_undetected_fraud_cases,
    (SELECT COUNT(*) FROM paysim WHERE isFraud = 1) AS total_fraud_cases,
    ROUND( COUNT(*) * 100.0 / (SELECT COUNT(*) FROM paysim WHERE isFraud = 1),4) AS undetected_fraud_percentage,
    CONCAT('KSH ', FORMAT(SUM(loss), 2)) AS total_undetected_fraud_value,
    CONCAT('KSH ', FORMAT(AVG(loss), 2)) AS avg_undetected_fraud_value
FROM missed_fraud;

 * mysql+pymysql://root:***@127.0.0.1:3306/fraud_analysis
1 rows affected.


total_undetected_fraud_cases,total_fraud_cases,undetected_fraud_percentage,total_undetected_fraud_value,avg_undetected_fraud_value
8197,8213,99.8052,"KSH 11,968,599,360.44","KSH 1,460,119.48"


### Insight
The system fails to detect 99.8% of fraudulent transactions, leaving 8,197 cases undetected with a total value of KSH 11.97 billion. This level of exposure is significantly higher than reported real-world losses, suggesting that actual fraud risk may be underreported or that the system is highly vulnerable under realistic conditions. The average loss of ~KSH 1.46M per case compounds into substantial cumulative damage when detection fails at scale. From a business perspective, this indicates a high level of uncontrolled financial risk, where fraud is able to pass through the system with minimal interruption.

## Q3: False Positive Rate and User Experience Impact
How frequently does the system incorrectly flag legitimate transactions as fraudulent, and what does this reveal about the trade-off between fraud prevention and user experience within the platform?

In [10]:
%%sql
SELECT
    COUNT(CASE WHEN isFraud = 0 THEN 1 END) AS total_legitimate_transactions,

    COUNT(CASE WHEN isFraud = 0 AND isFlaggedFraud = 1 THEN 1 END) AS false_positives,

    ROUND(
        COUNT(CASE WHEN isFraud = 0 AND isFlaggedFraud = 1 THEN 1 END) * 100.0 /
        COUNT(CASE WHEN isFraud = 0 THEN 1 END),
        6
    ) AS false_positive_rate_percentage
FROM paysim;

 * mysql+pymysql://root:***@127.0.0.1:3306/fraud_analysis
1 rows affected.


total_legitimate_transactions,false_positives,false_positive_rate_percentage
6354407,0,0.00000


### Insight

The system records zero false positives, meaning legitimate transactions are never incorrectly flagged, which suggests a frictionless user experience. However, when viewed alongside the 99.8% missed fraud rate, this does not indicate strong performance it indicates that the system is barely engaging in fraud detection at all.
In practical terms, the system is positioned at the extreme end of the trade-off spectrum: it fully protects user experience but fails to provide meaningful fraud prevention. A detection system that produces no false positives but also fails to detect almost all fraud is not effective, it is functionally weak, as it allows fraudulent activity to pass through without resistance while maintaining the appearance of accuracy.
